overlord/snapstate/handlers.go

```diff
@@ -3532,17 +3637,80 @@ func (m *SnapManager) doSetupAliases(t *state.Task, _ *tomb.Tomb) error {
 	}
 	snapName := snapsup.InstanceName()
 	curAliases := snapst.Aliases
+	autoDisabled := snapst.AutoAliasesDisabled
+
+	var oldAliases map[string]*AliasTarget
+	oldAutoDisabled := autoDis
 
-	_, _, err = applyAliasesChange(snapName, autoDis, nil, snapst.AutoAliasesDisabled, curAliases, m.backend, doApply)
+	prune, err := shouldPruneOldAliases(t, snapName)
 	if err != nil {
 		return err
 	}
+	if prune {
+		oldAliases, err = snapOldAliases(t, snapName)
+		if err != nil {
+			return err
+		}
+		oldAutoDisabled = snapst.AutoAliasesDisabled
+	}
 
+	// no need to check for conflicts as it was alredy checked in `set-auto-aliases`
+	_, _, err = applyAliasesChange(snapName, oldAutoDisabled, oldAliases, autoDisabled, curAliases, m.backend, doApply)
+	if err != nil {
+		return err
+	}
+
+	t.Set("old-aliases-v2", oldAliases)
+	t.Set("old-auto-aliases-disabled", oldAutoDisabled)
 	snapst.AliasesPending = false
 	Set(st, snapName, snapst)
 	return nil
 }
 
+func (m *SnapManager) undoSetupAliases(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+	snapsup, snapst, err := snapSetupAndState(t)
+	if err != nil {
+		return err
+	}
+	snapName := snapsup.InstanceName()
+
+	if !snapst.AliasesPending {
+		curAliases := snapst.Aliases
+		curAutoDisabled := snapst.AutoAliasesDisabled
+		var oldAliases map[string]*AliasTarget
+		if err := t.Get("old-aliases-v2", &oldAliases); err != nil && !errors.Is(err, state.ErrNoState) {
+			return err
+		}
+		oldAutoDisabled := autoDis
+		if err := t.Get("old-auto-aliases-disabled", &oldAutoDisabled); err != nil && !errors.Is(err, state.ErrNoState) {
+			return err
+		}
+
+		// check if the old states creates conflicts now
+		_, err = checkAliasesConflicts(st, snapName, oldAutoDisabled, oldAliases, nil)
+		if _, ok := err.(*AliasConflictError); ok {
+			// best we can do is reinstate with all aliases disabled
+			t.Errorf("cannot reinstate alias state because of conflicts, disabling: %v", err)
+			oldAliases, _ = disableAliases(oldAliases)
+			oldAutoDisabled = true
+			snapst.AutoAliasesDisabled = true
+		} else if err != nil {
+			return err
+		}
+
+		if _, _, err := applyAliasesChange(snapName, curAutoDisabled, curAliases, oldAutoDisabled, oldAliases, m.backend, doApply); err != nil {
+			return err
+		}
```

Comment:
this code looks very similar to undoRefreshAliases? which is the undo of set-auto-aliases as well, so I'm not fully sure we need this?

Reply:
or we want this then maybe we can simplify the undo of set-auto-aliases instead?

Reply:
That's a great idea actually, a separate undo for `set-auto-aliases` would be much simple and straightforward.

Reply:
An alternative approach to avoid further code duplication is adding small logic into `undoRefreshAliases` to adapt it as an undo for `setup-aliases`, only doing changes on disk and not affecting state aliases?

Reply:
of course that would make `undoRefreshAliases` more complex, but much less logic duplication.